pkg/workload/workload.go

```diff
@@ -39,6 +40,40 @@ var (
 	admissionManagedConditions = []string{kueue.WorkloadAdmitted, kueue.WorkloadEvicted}
 )
 
+type FlavorResourceQuantities map[kueue.ResourceFlavorReference]map[corev1.ResourceName]int64
+type LastScheduleClusterQueueState struct {
+	LastScheduledFlavorIdx map[string]map[corev1.ResourceName]int
+	ResourceFlavors        map[kueue.ResourceFlavorReference]*kueue.ResourceFlavor
+	ClusterQueueUsage      FlavorResourceQuantities
+	CohortUsage            FlavorResourceQuantities
+}
+
+func (s *LastScheduleClusterQueueState) Equal(resourceFlavors map[kueue.ResourceFlavorReference]*kueue.ResourceFlavor, clusterQueueUsage, cohortUsage FlavorResourceQuantities) bool {
+	return reflect.DeepEqual(s.ResourceFlavors, resourceFlavors) && reflect.DeepEqual(s.ClusterQueueUsage, clusterQueueUsage) && reflect.DeepEqual(s.CohortUsage, cohortUsage)
+}
+
+func (s *LastScheduleClusterQueueState) Clone() *LastScheduleClusterQueueState {
+	c := LastScheduleClusterQueueState{
+		LastScheduledFlavorIdx: make(map[string]map[corev1.ResourceName]int),
+		ResourceFlavors:        make(map[kueue.ResourceFlavorReference]*kueue.ResourceFlavor),
+		ClusterQueueUsage:      make(FlavorResourceQuantities),
+		CohortUsage:            make(FlavorResourceQuantities),
+	}
+	for ps, flavorIdx := range s.LastScheduledFlavorIdx {
+		c.LastScheduledFlavorIdx[ps] = maps.Clone(flavorIdx)
+	}
+	for res, flavor := range s.ResourceFlavors {
+		c.ResourceFlavors[res] = flavor
+	}
+	for flavor, flavorusage := range s.ClusterQueueUsage {
+		c.ClusterQueueUsage[flavor] = maps.Clone(flavorusage)
+	}
+	for flavor, flavorusage := range s.CohortUsage {
+		c.CohortUsage[flavor] = maps.Clone(flavorusage)
+	}
+	return &c
+}
+
 // Info holds a Workload object and some pre-processing.
 type Info struct {
```

Comment:
Not sure if this will be relevant, but worth checking.

When a admission attempt fails, we usually send a status update to the workload with an Admitted=False condition

https://github.com/kubernetes-sigs/kueue/blob/3d1dc9a4fb78c52bdb8ca53d8c94139a6f241304/pkg/scheduler/scheduler.go#L470-L471

The update to the object would cause an update into the queue:

https://github.com/kubernetes-sigs/kueue/blob/3d1dc9a4fb78c52bdb8ca53d8c94139a6f241304/pkg/controller/core/workload_controller.go#L297

The problem is that in there we always call `workload.NewInfo`, which could wipe the saved status.

Note that requeuing happens before sending the update https://github.com/kubernetes-sigs/kueue/blob/3d1dc9a4fb78c52bdb8ca53d8c94139a6f241304/pkg/scheduler/scheduler.go#L466-L471

But with enough retries, it is possible that a Workload update comes while we are doing a scheduling attempt. Then, we might need to change `requeueIfNotPresent` to still update the lastState.

https://github.com/kubernetes-sigs/kueue/blob/3d1dc9a4fb78c52bdb8ca53d8c94139a6f241304/pkg/queue/cluster_queue_impl.go#L130

Reply:
I will add some check in `Manager.addOrUpdateWorkload()` to judge whether to copy lastState